exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/TypeInferenceUtils.java

```diff
@@ -0,0 +1,568 @@
+import org.apache.drill.exec.resolver.FunctionResolverFactory;
+import org.apache.drill.exec.resolver.TypeCastRules;
+
+import java.util.List;
+import java.util.Map;
+
+public class TypeInferenceUtils {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(TypeInferenceUtils.class);
+
+  public static final TypeProtos.MajorType UNKNOWN_TYPE = TypeProtos.MajorType.getDefaultInstance();
+  private static ImmutableMap<TypeProtos.MinorType, SqlTypeName> DRILL_TO_CALCITE_TYPE_MAPPING =
+      ImmutableMap.<TypeProtos.MinorType, SqlTypeName> builder()
+          .put(TypeProtos.MinorType.INT, SqlTypeName.INTEGER)
+          .put(TypeProtos.MinorType.BIGINT, SqlTypeName.BIGINT)
+          .put(TypeProtos.MinorType.FLOAT4, SqlTypeName.FLOAT)
+          .put(TypeProtos.MinorType.FLOAT8, SqlTypeName.DOUBLE)
+          .put(TypeProtos.MinorType.VARCHAR, SqlTypeName.VARCHAR)
+          .put(TypeProtos.MinorType.BIT, SqlTypeName.BOOLEAN)
+          .put(TypeProtos.MinorType.DATE, SqlTypeName.DATE)
+          .put(TypeProtos.MinorType.DECIMAL9, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.DECIMAL18, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.DECIMAL28SPARSE, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.DECIMAL38SPARSE, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.TIME, SqlTypeName.TIME)
+          .put(TypeProtos.MinorType.TIMESTAMP, SqlTypeName.TIMESTAMP)
+          .put(TypeProtos.MinorType.VARBINARY, SqlTypeName.VARBINARY)
+          .put(TypeProtos.MinorType.INTERVALYEAR, SqlTypeName.INTERVAL_YEAR_MONTH)
+          .put(TypeProtos.MinorType.INTERVALDAY, SqlTypeName.INTERVAL_DAY_TIME)
+          .put(TypeProtos.MinorType.MAP, SqlTypeName.MAP)
+          .put(TypeProtos.MinorType.LIST, SqlTypeName.ARRAY)
+          .put(TypeProtos.MinorType.LATE, SqlTypeName.ANY)
+          .build();
+
+  private static ImmutableMap<SqlTypeName, TypeProtos.MinorType> CALCITE_TO_DRILL_MAPPING =
+      ImmutableMap.<SqlTypeName, TypeProtos.MinorType> builder()
+          .put(SqlTypeName.INTEGER, TypeProtos.MinorType.INT)
+          .put(SqlTypeName.BIGINT, TypeProtos.MinorType.BIGINT)
+          .put(SqlTypeName.FLOAT, TypeProtos.MinorType.FLOAT4)
+          .put(SqlTypeName.DOUBLE, TypeProtos.MinorType.FLOAT8)
+          .put(SqlTypeName.VARCHAR, TypeProtos.MinorType.VARCHAR)
+          .put(SqlTypeName.BOOLEAN, TypeProtos.MinorType.BIT)
+          .put(SqlTypeName.DATE, TypeProtos.MinorType.DATE)
+          .put(SqlTypeName.TIME, TypeProtos.MinorType.TIME)
+          .put(SqlTypeName.TIMESTAMP, TypeProtos.MinorType.TIMESTAMP)
+          .put(SqlTypeName.VARBINARY, TypeProtos.MinorType.VARBINARY)
+          .put(SqlTypeName.INTERVAL_YEAR_MONTH, TypeProtos.MinorType.INTERVALYEAR)
+          .put(SqlTypeName.INTERVAL_DAY_TIME, TypeProtos.MinorType.INTERVALDAY)
+          .put(SqlTypeName.CHAR, TypeProtos.MinorType.VARCHAR)
+          .put(SqlTypeName.DECIMAL, TypeProtos.MinorType.FLOAT8)
```

Comment:
Please add a comment to:
SqlTypeName.CHAR -> TypeProtos.MinorType.VARCHAR

This one is incorrect?
SqlTypeName.DECIMAL -> TypeProtos.MinorType.FLOAT8

Also, aren't there more types here too?